File: client/lib/archive/codec.ts
/**
 * Encode and decode PortfolioDB archive documents.
 *
 * The encoding decisions are made here and only here, and mirror
 * server/archive/codec.go: an archive written by one runtime has to be the same
 * bytes as one written by the other. See docs/spec/archive-format.md.
 *
 * React-free, so the browser extension can import it through its path alias.
 */

import { create, fromJson, toJsonString } from "@bufbuild/protobuf";
import type { DescEnum, DescMessage, MessageInitShape, MessageShape } from "@bufbuild/protobuf";
import {
  SystemArchiveSchema,
  UserArchiveSchema,
} from "@/gen/archive/v1/archive_pb";
import type { SystemArchive, UserArchive } from "@/gen/archive/v1/archive_pb";
import { ArchiveKind } from "@/gen/archive/v1/common_pb";

/** What this build writes, and the highest it will read. */
export const FORMAT_VERSION = 1;

const writeOpts = {
  // snake_case keys. Without this protobuf-es writes lowerCamelCase, which is a
  // total format change and a silent one.
  useProtoFieldName: true,
  // Enums by name: "STOCK", not 1.
  enumAsInteger: false,
  // Absent stays absent, which is the whole point of the optional fields.
  alwaysEmitImplicit: false,
} as const;

const readOpts = {
  // A field added by a later PortfolioDB is dropped rather than failing the
  // parse. Safe only because the format version is checked first.
  //
  // In protobuf-es this flag also swallows an unrecognised enum *value* name,
  // leaving the field at zero, where Go's protojson rejects it. That divergence
  // is not acceptable here -- a broker read as unspecified is a wrong value
  // rather than an error -- so assertKnownEnums restores the stricter
  // behaviour and the two runtimes agree.
  ignoreUnknownFields: true,
} as const;

/** An archive written by a later PortfolioDB than this build understands. */
export class ArchiveVersionError extends Error {
  constructor(
    readonly got: number,
    readonly supports: number
  ) {
    super(`archive format version ${got} is newer than this client supports (${supports})`);
    this.name = "ArchiveVersionError";
  }
}

/** A user archive handed to the system importer, or the reverse. */
export class ArchiveKindError extends Error {
  constructor(
    readonly got: string,
    readonly want: string
  ) {
    super(`archive is ${got}, expected ${want}`);
    this.name = "ArchiveKindError";
  }
}

/**
 * A sentence a user can act on, for an error thrown by unmarshalSystem or
 * unmarshalUser.
 *
 * Whether a file is a valid archive is an all-or-nothing question answered at
 * parse time, so this is the only thing an upload UI has to say about a
 * rejected file. Whether its contents resolve against this instance is a
 * separate, per-group question, answered later by the job's validation errors.
 */
export function archiveErrorMessage(e: unknown): string {
  if (e instanceof ArchiveVersionError) {
    return `This file was written by a later PortfolioDB (format version ${e.got}). Upgrade this instance to read it.`;
  }
  if (e instanceof ArchiveKindError) {
    return `This is a ${e.got.toLowerCase()} archive, not a ${e.want.toLowerCase()} one.`;
  }
  return e instanceof Error ? e.message : String(e);
}

/** Write a system archive. The envelope's version and kind are stamped here. */
export function marshalSystem(archive: MessageInitShape<typeof SystemArchiveSchema>): string {
  return marshal(SystemArchiveSchema, archive, ArchiveKind.SYSTEM);
}

/** Write a user archive. */
export function marshalUser(archive: MessageInitShape<typeof UserArchiveSchema>): string {
  return marshal(UserArchiveSchema, archive, ArchiveKind.USER);
}

/** Read a system archive. Throws ArchiveVersionError or ArchiveKindError. */
export function unmarshalSystem(json: string): SystemArchive {
  return unmarshal(SystemArchiveSchema, json, ArchiveKind.SYSTEM);
}

/** Read a user archive. */
export function unmarshalUser(json: string): UserArchive {
  return unmarshal(UserArchiveSchema, json, ArchiveKind.USER);
}

function marshal<Desc extends DescMessage>(
  schema: Desc,
  init: MessageInitShape<Desc>,
  kind: ArchiveKind
): string {
  const msg = create(schema, init) as MessageShape<Desc> & {
    envelope?: { formatVersion: number; kind: ArchiveKind };
  };
  // Stamped rather than taken from the caller, so no call site can write a
  // document that misdescribes its own version or kind. Created when absent, so
  // a caller cannot write one with no envelope at all either.
  const envelopeField = schema.fields.find((f) => f.name === "envelope");
  if (!msg.envelope && envelopeField?.fieldKind === "message") {
    (msg as { envelope?: unknown }).envelope = create(envelopeField.message);
  }
  const envelope = msg.envelope as { formatVersion: number; kind: ArchiveKind };
  envelope.formatVersion = FORMAT_VERSION;
  envelope.kind = kind;
  return toJsonString(schema, msg, writeOpts);
}

// The document is parsed once and the parsed value is reused: the envelope
// probe, the enum check and the message decode would otherwise each walk it,
// and a price archive is megabytes.
function unmarshal<Desc extends DescMessage>(
  schema: Desc,
  json: string,
  kind: ArchiveKind
): MessageShape<Desc> {
  let doc: unknown;
  try {
    doc = JSON.parse(json);
  } catch (e) {
    throw new Error(`not an archive document: ${e instanceof Error ? e.message : String(e)}`);
  }
  probe(doc, kind);
  assertKnownEnums(schema, doc, "");
  return fromJson(schema, doc as never, readOpts);
}

/**
 * Check the envelope before decoding the document, because decoding is what a
 * version mismatch would break: a reader that failed part-way would report a
 * confusing parse error where it should report that the file was written by a
 * later PortfolioDB.
 *
 * Both spellings of the key are accepted: protobuf-es writes snake_case and
 * reads either, so a hand-written document using lowerCamelCase parses.
 */
function probe(doc: unknown, want: ArchiveKind): void {
  const envelope = (doc as { envelope?: Record<string, unknown> } | null)?.envelope;
  if (!envelope) throw new Error("archive has no envelope");

  const raw = envelope.format_version ?? envelope.formatVersion;
  const version = typeof raw === "number" ? raw : Number(raw);
  if (!version || Number.isNaN(version)) throw new Error("archive has no format_version");
  if (version > FORMAT_VERSION) throw new ArchiveVersionError(version, FORMAT_VERSION);

  const kind = envelope.kind;
  if (typeof kind !== "string" || !(kind in ArchiveKind)) {
    throw new Error(`archive kind ${JSON.stringify(kind)} is not a known kind`);
  }
  if (ArchiveKind[kind as keyof typeof ArchiveKind] !== want) {
    throw new ArchiveKindError(kind, ArchiveKind[want]);
  }
}

/**
 * Walk a parsed document against its descriptor and reject any enum value the
 * vocabulary does not define. A name outside the vocabulary is not
 * representable anywhere in PortfolioDB, not merely in this file, so it is an
 * error rather than something to drop -- and dropping it would leave a zero,
 * which reads as a different value rather than as nothing.
 *
 * Unknown *fields* are still ignored; only unknown enum values are refused.
 */
function assertKnownEnums(schema: DescMessage, value: unknown, path: string): void {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return;
  const obj = value as Record<string, unknown>;

  for (const field of schema.fields) {
    const raw = obj[field.name] ?? obj[field.jsonName];
    if (raw === undefined || raw === null) continue;
    const at = path ? `${path}.${field.name}` : field.name;

    switch (field.fieldKind) {
      case "enum":
        checkEnum(field.enum, raw, at);
        break;
      case "message":
        assertKnownEnums(field.message, raw, at);
        break;
      case "list":
        if (!Array.isArray(raw)) break;
        raw.forEach((item, i) => {
          if (field.listKind === "enum") checkEnum(field.enum, item, `${at}[${i}]`);
          else if (field.listKind === "message") assertKnownEnums(field.message, item, `${at}[${i}]`);
        });
        break;
      case "map":
        for (const [k, v] of Object.entries(raw as Record<string, unknown>)) {
          if (field.mapKind === "enum") checkEnum(field.enum, v, `${at}[${k}]`);
          else if (field.mapKind === "message") assertKnownEnums(field.message, v, `${at}[${k}]`);
        }
        break;
    }
  }
}

function checkEnum(desc: DescEnum, value: unknown, at: string): void {
  // protojson accepts a value by number as well as by name.
  if (typeof value === "number") return;
  if (typeof value !== "string" || !desc.values.some((v) => v.name === value)) {
    throw new Error(`${at}: ${JSON.stringify(value)} is not a known ${desc.name}`);
  }
}

File: server/archive/codec.go
// Package archive encodes and decodes PortfolioDB archive documents.
//
// The encoding decisions are made here and only here. An archive is written
// today and read by a server that does not exist yet, so the options that
// decide what it looks like on disk -- proto field names, enums by name, absent
// fields staying absent -- cannot be restated at each call site: one caller
// forgetting UseProtoNames silently rewrites every key in the file.
//
// See docs/spec/archive-format.md.
package archive

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"

	"buf.build/go/protovalidate"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"

	archivev1 "github.com/leedenison/portfoliodb/proto/archive/v1"
)

// FormatVersion is what this build writes, and the highest it will read.
//
// 2: a listing is named by a security identifier and a currency, so InstrumentRef
// carries one and PriceGroup's own currency field is gone -- a field removed and
// a message renumbered, which is what the version exists to catch.
const FormatVersion = 2

var marshalOpts = protojson.MarshalOptions{
	// snake_case keys. Without this protojson writes lowerCamelCase, which is a
	// total format change and a silent one.
	UseProtoNames: true,
	// Enums by name: "STOCK", not 1.
	UseEnumNumbers: false,
	// Absent stays absent, which is the whole point of the optional fields.
	EmitUnpopulated: false,
	AllowPartial:    false,
}

var unmarshalOpts = protojson.UnmarshalOptions{
	// A field added by a later PortfolioDB is dropped rather than failing the
	// parse. That is safe only because format_version is checked first: an
	// additive change leaves the version alone, and a change an older reader
	// cannot survive bumps it and is refused outright.
	DiscardUnknown: true,
	AllowPartial:   false,
}

// VersionError reports an archive written by a later PortfolioDB.
type VersionError struct {
	Got      uint32
	Supports uint32
}

func (e *VersionError) Error() string {
	return fmt.Sprintf("archive format version %d is newer than this server supports (%d): upgrade to read it", e.Got, e.Supports)
}

// KindError reports an archive handed to the wrong importer -- a user archive
// given to the system path, or the reverse.
type KindError struct {
	Got  archivev1.ArchiveKind
	Want archivev1.ArchiveKind
}

func (e *KindError) Error() string {
	return fmt.Sprintf("archive is %s, expected %s", e.Got, e.Want)
}

// NewEnvelope builds the envelope an export stamps onto its own document. The
// format version and the kind come from this build rather than from a caller,
// for the same reason MarshalSystem overwrites them.
//
// exported_at is knowledge time -- when the data being exported was current --
// so it is the exporting server's clock and nothing else's.
func NewEnvelope(sourceInstance string, kind archivev1.ArchiveKind) *archivev1.Envelope {
	return &archivev1.Envelope{
		FormatVersion:  FormatVersion,
		ExportedAt:     timestamppb.Now(),
		SourceInstance: sourceInstance,
		Kind:           kind,
	}
}

// CheckEnvelope validates an envelope that arrived over the API rather than
// inside a file. UnmarshalSystem probes the bytes before parsing them, but an
// RPC carrying an already-parsed envelope has no bytes to probe, and the two
// refusals -- a file from a later PortfolioDB, a file handed to the wrong
// importer -- have to happen either way.
func CheckEnvelope(e *archivev1.Envelope, want archivev1.ArchiveKind) error {
	if e == nil {
		return errors.New("archive has no envelope")
	}
	if e.GetFormatVersion() == 0 {
		return errors.New("archive has no format_version")
	}
	if e.GetFormatVersion() > FormatVersion {
		return &VersionError{Got: e.GetFormatVersion(), Supports: FormatVersion}
	}
	if e.GetKind() != want {
		return &KindError{Got: e.GetKind(), Want: want}
	}
	return nil
}

// MarshalSystem writes a system archive. It stamps the envelope's format_version
// and kind itself, so a caller cannot write a document that misdescribes itself.
func MarshalSystem(a *archivev1.SystemArchive) ([]byte, error) {
	a = proto.CloneOf(a)
	if a.Envelope == nil {
		a.Envelope = &archivev1.Envelope{}
	}
	a.Envelope.FormatVersion = FormatVersion
	a.Envelope.Kind = archivev1.ArchiveKind_SYSTEM
	return marshal(a)
}

// MarshalUser writes a user archive.
func MarshalUser(a *archivev1.UserArchive) ([]byte, error) {
	a = proto.CloneOf(a)
	if a.Envelope == nil {
		a.Envelope = &archivev1.Envelope{}
	}
	a.Envelope.FormatVersion = FormatVersion
	a.Envelope.Kind = archivev1.ArchiveKind_USER
	return marshal(a)
}

// UnmarshalSystem reads a system archive.
func UnmarshalSystem(b []byte) (*archivev1.SystemArchive, error) {
	if err := probe(b, archivev1.ArchiveKind_SYSTEM); err != nil {
		return nil, err
	}
	var a archivev1.SystemArchive
	if err := unmarshal(b, &a); err != nil {
		return nil, err
	}
	return &a, nil
}

// UnmarshalUser reads a user archive.
func UnmarshalUser(b []byte) (*archivev1.UserArchive, error) {
	if err := probe(b, archivev1.ArchiveKind_USER); err != nil {
		return nil, err
	}
	var a archivev1.UserArchive
	if err := unmarshal(b, &a); err != nil {
		return nil, err
	}
	return &a, nil
}

func marshal(m proto.Message) ([]byte, error) {
	if err := protovalidate.Validate(m); err != nil {
		return nil, fmt.Errorf("invalid archive: %w", err)
	}
	b, err := marshalOpts.Marshal(m)
	if err != nil {
		return nil, fmt.Errorf("marshal archive: %w", err)
	}
	// protojson injects unstable whitespace on purpose, so that callers cannot
	// depend on its exact bytes. An archive is a file a user diffs and a script
	// writes, so normalise it: two exports of the same data are then identical.
	var out bytes.Buffer
	if err := json.Compact(&out, b); err != nil {
		return nil, fmt.Errorf("compact archive: %w", err)
	}
	return out.Bytes(), nil
}

func unmarshal(b []byte, m proto.Message) error {
	if err := unmarshalOpts.Unmarshal(b, m); err != nil {
		return fmt.Errorf("parse archive: %w", err)
	}
	if err := protovalidate.Validate(m); err != nil {
		return fmt.Errorf("invalid archive: %w", err)
	}
	return nil
}

// probe reads the envelope with encoding/json before the real parse, because
// the real parse is what a version mismatch would break: a reader that failed
// on an unknown field would report a confusing parse error where it should
// report that the file was written by a later PortfolioDB.
//
// It accepts both spellings of the key. protojson writes snake_case and reads
// either, so a hand-written document using lowerCamelCase parses.
func probe(b []byte, want archivev1.ArchiveKind) error {
	var p struct {
		Envelope struct {
			FormatVersion  uint32 `json:"format_version"`
			FormatVersionC uint32 `json:"formatVersion"`
			Kind           string `json:"kind"`
		} `json:"envelope"`
	}
	if err := json.Unmarshal(b, &p); err != nil {
		return fmt.Errorf("not an archive document: %w", err)
	}
	v := p.Envelope.FormatVersion
	if v == 0 {
		v = p.Envelope.FormatVersionC
	}
	if v == 0 {
		return errors.New("archive has no format_version")
	}
	if v > FormatVersion {
		return &VersionError{Got: v, Supports: FormatVersion}
	}
	got, ok := archivev1.ArchiveKind_value[p.Envelope.Kind]
	if !ok {
		return fmt.Errorf("archive kind %q is not a known kind", p.Envelope.Kind)
	}
	if archivev1.ArchiveKind(got) != want {
		return &KindError{Got: archivev1.ArchiveKind(got), Want: want}
	}
	return nil
}
